framework/Renderers/Source/Renderer.swift

```diff
@@ -16,5 +16,6 @@ public protocol Renderer: AnyObject{
 	func drawSolidRectWithBorder(topLeftPoint p1: Point,topRightPoint p2: Point,bottomRightPoint p3: Point,bottomLeftPoint p4: Point, strokeWidth thickness: Float, fillColor: Color, borderColor: Color)
 	func getTextWidth(text: String, textSize size: Float) -> Float
 	func drawOutput(fileName name: String)
+	func base64Png(fileName name : String) -> String
```

Comment:
I suggest not adding this as a function in the protocol, so that you don't have to define implementations in renderers that do not support it, and so that users can statically see (e.g. in the autocomplete suggestions for a renderer) whether it supports it.

**Extra thoughts. No need to do any of these in this PR:**

Actually, I have the same opinion about `drawOutput` -- it shouldn't be in the protocol, so that renderers that don't support saving output to a file don't have to implement it.

Of course, if you do this, then it won't be possible to have methods like `LinePlot.drawGraphAndOutput(filename: String, renderer: Renderer)`, because that function won't have access to a `renderer.drawOutput` method. I don't think this is a big problem -- for now, the user can just do:
```
let renderer = ...
let plot = ...
plot.drawGraph(renderer: renderer)
renderer.drawOutput(filename: "abc")
```

Then later, there are ways to improve the api so that the user doesn't have to make so many calls to draw a graph and save it to a file. One idea: make a protocol for plots
```
protocol Plot: {
  func draw(on renderer: Renderer)
}
```
Then, it's possible to write functions that "draw and save" plots using renderers. e.g.
```
// maybe this is a top-level function inside swiftplot
func saveAGGPng(for plot: Plot, to filename: String) {
  let renderer = AGGRenderer()
  plot.draw(on: renderer)
  renderer.save(to: filename)
}

// maybe this is a function inside EnableIPythonDisplay in swift-jupyter
func display(plot: Plot) {
  let renderer = AGGRenderer()
  plot.draw(on: renderer)
  let base64Png = renderer.base64Png()
  ... logic to show the png on jupyter ...
}
```

Reply:
ooo even better, those "saveAGGPng" functions could be methods in a protocol extension:
```
extension Plot {
  func saveAGGPng(to filename: String) { ... }
}

extension Plot {
  func display() { ... }
}
```

so that the user can just do

```
let plot = ...
plot.saveAGGPng(to: "abc")
plot.display()
```

Reply:
Regarding the `drawOutput` function, I don't think the name indicates that the renderer need to save the image. So maybe if we have OpenGL the code inside this method can draw the output to a window on the screen?

Reply:
The argument indicates that it's a function to save an image though, and what if other renderers need different types of arguments to draw their output (e.g. something that draws to a screen needs a screen position)?

Stepping back a bit, I think that the concepts of passing configuration to the renderer (e.g. the filename to save to) and deciding how to receive the image (e.g. as a saved file or as a returned string), should be part of the specific renderer implementation, not part of the abstract renderer interface. This way, renderer implementations have complete freedom to define how exactly they are going to draw their plots, without being constrained by existing methods in the renderer interface.